boa/src/builtins/object/gcobject.rs

```diff
@@ -61,6 +65,55 @@ impl Display for BorrowError {
     }
 }
 
+struct RecursionLimiter {
+    free: bool,
+    first: bool,
+}
+
+impl RecursionLimiter {
+    thread_local! {
+        pub static VISITED: RefCell<HashSet<usize>> = RefCell::new(HashSet::new());
+    }
+
+    fn new(o: &GcObject) -> Self {
+        // We shouldn't have to worry too much about this being moved during Debug::fmt.
+        let ptr = (&*o.borrow() as *const Object) as usize;
+        let (free, first) = Self::VISITED.with(|hs| {
+            let mut hs = hs.borrow_mut();
+            (hs.len() == 0, hs.insert(ptr))
+        });
+
+        Self { free, first }
+    }
+}
+
+impl Drop for RecursionLimiter {
+    fn drop(&mut self) {
+        // Typically, calling hs.remove(ptr) for "first" objects would be the correct choice here. This would allow the
+        // same object to appear multiple times in the output (provided it does not appear under itself recursively).
+        // However, the JS object hierarchy involves quite a bit of repitition, and the sheer amount of data makes
+        // understanding the Debug output impossible; limiting the usefulness of it.
+        //
+        // Instead, the entire hashset is emptied at by the first GcObject involved. This means that objects will appear
+        // at most once, throughout the graph, hopefully making things a bit clearer.
+        if self.free {
+            Self::VISITED.with(|hs| hs.borrow_mut().clear());
+        }
+    }
+}
+
+impl Debug for GcObject {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        let limiter = RecursionLimiter::new(&self);
+
+        if limiter.first {
+            f.debug_tuple("GcObject").field(&self.0).finish()
+        } else {
+            f.write_str("...")
```

Comment:
I think this makes it clear this is an object:
```suggestion
            f.write_str("{ ... }")
```